src/input/MeshAttributes.cpp

```diff
@@ -38,250 +35,123 @@ void MeshAttributes::readXmlFile(const char* xmlFilename) {
-       for(int i = 0; i < tokens.size(); i++) {
-           int faceid = std::atoi(tokens[i].c_str());
-           lFaceIdMeshSizePropagation.push_back(faceid);
-       }
+    retrieve_attribute_and_update_class_member(&doc, "MeshSizePropagation", "ScalingFactor", &MeshSizePropagationScalingFactor);
+    retrieve_attribute_and_update_class_member(&doc, "MeshSizePropagation", "Distance", &MeshSizePropagationDistance);
+    if (auto child = doc.FirstChildElement("MeshSizePropagation")) {
+       lFaceIdMeshSizePropagation = parse_string_and_fill_list(child-> GetText());
     }
 }
 
 
 void MeshAttributes::set_regionMSize() {
-    std::string line,sval;
-    XMLElement* child;
-    XMLElement* pRoot;
-    child = doc.FirstChildElement("regionMSize");
-    for (child; child; child = child->NextSiblingElement("regionMSize"))
+    for (auto child = doc.FirstChildElement("regionMSize"); child; child = child->NextSiblingElement("regionMSize"))
     {
-       sval = child->Attribute("value");
-       double regionMSize =  atof(sval.c_str());
-       lregionMSize.push_back(regionMSize);
-       std::vector<std::string> tokens;
-       line =  child-> GetText();
-       split(tokens, line, ',');
-       std::list<int> lregionId;
-       for(int i = 0; i < tokens.size(); i++) {
-           int regionid = std::atoi(tokens[i].c_str());
-           lregionId.push_back(regionid);
-       }
-       llregionMSizeRegionId.push_back(lregionId);
+       lregionMSize.push_back( std::atof(child->Attribute("value")));
+       llregionMSizeRegionId.push_back(parse_string_and_fill_list(child-> GetText()));
     }
 }
 
 
-void MeshAttributes::set_gradation() {
-    std::string sval;
-    XMLElement* pRoot;
-    pRoot = doc.FirstChildElement("gradation");
-    if(pRoot) {
-       sval = pRoot->Attribute("value");
-       gradation =  atof(sval.c_str());
-    }
-}
+void MeshAttributes::set_global_mesh_attributes() {
+    retrieve_attribute_and_update_class_member(&doc, "gradation", "value", &gradation);
+    retrieve_attribute_and_update_class_member(&doc, "globalMSize", "value", &globalMSize);
+    retrieve_attribute_and_update_class_member(&doc, "area_AspectRatio", "value", &area_AspectRatio);
+    retrieve_attribute_and_update_class_member(&doc, "vol_AspectRatio", "value", &vol_AspectRatio);
 
-void MeshAttributes::set_globalMSize() {
-    std::string sval;
-    XMLElement* pRoot;
-    pRoot = doc.FirstChildElement("globalMSize");
-    if(pRoot) {
-       sval = pRoot->Attribute("value");
-       globalMSize =  atof(sval.c_str());
-    }
 }
 
 
-void MeshAttributes::set_area_AspectRatio() {
-    std::string sval;
-    XMLElement* pRoot;
-    pRoot = doc.FirstChildElement("area_AspectRatio");
-    if(pRoot) {
-       sval = pRoot->Attribute("value");
-       area_AspectRatio =  atof(sval.c_str());
+void MeshAttributes::retrieve_attribute_and_update_class_member(XMLNode* element, const char* sElementName, const char* sAttributeName, double* MeshAttributesMember) {
```

Comment:
You could make `element` a reference. Also https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Rf-logical this rule states, that a function should not have two logical purposes.
What about e.g. `gradation = retrieve_attribute(doc, "gradation", "value")`

Reply:
if you do ```gradation = retrieve_attribute(doc, "gradation", "value")```, then gradation get updated even if element has no child with name ```sElementName```, that why I merged these 2 steps in one function.

Reply:
If you call it `update_attribute_from_xml` there's no and in the function name any more, so automatic semantic checkers don't see this as an issue any more ;-)

Reply:
```
void MeshAttributes::update_attribute_from_xml(XMLNode& element, const char* sElementName, const char* sAttributeName, double& MeshAttributesMember) {
    if (auto child = element->FirstChildElement(sElementName)) {
        const char* attr = child->Attribute(sAttributeName);
        if (attr != null) {
            MeshAttributesMember = std::atof(attr);
        } else {
            logError() << XML Tag " << sElementName << " has no attribute " << sAttributeName;
        }
    }
}
```

Reply:
Thx. But actually, after some thoughts, the attr==null should not raise an error.
If the attribute is not specified it is not an error but a choice.
(e.g. in the GUI, one may choose to set a refinement zone (e.g. cube), or not).

Reply:
ok that's fine!